Fixes #32880 - Add confirmation modal service

Ron-Lavi opened the pull request:
The confirmation modal will be placed on the app's root
and can be triggered from everywhere. e.g: `openConfirmModal({ title, message, onConfirm })`

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/users/PersonalAccessTokens/PersonalAccessTokens.js:
```diff
@@ -32,10 +33,18 @@ const PersonalAccessTokens = ({ url, canCreate }) => {
+        label: __('Revoke'),
+        message: __('Do you really want to revoke Access Token?'),
+        onConfirm: closeCB => {
+          dispatch(revokePersonalAccessTokenAction({ url, id }));
```
@sufanek1 hey, what do you think about this approach?

sufanek1 replied:
why not :)

ofedoren commented:
@laviro, since it's a generic component, maybe it's worth to add to the stories, probably with an example?

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/ConfirmModal/ConfirmModal.js:
```diff
@@ -0,0 +1,71 @@
+  const closeModal = () => dispatch(closeConfirmModal());
+  const handleConfirm = () => {
+    setLoading(true);
+    onConfirm(closeModal);
```
why is the closeModal passed as argument?

Ron-Lavi replied:
I was thinking to let the consumer decide whether to close it based on an API response for example,
but I still want to play with that and see if we have a strong use case for that

ezr-ondrej replied:
IMHO if you do not have use-case in mind, it's better not to do it, simplier code always wins for me.
If you need it in the future, you can always add that :)

Ron-Lavi replied:
right, I agree with that!

ezr-ondrej reviewed (dismissed):
Looks quite nice!
Do we want to add an option to choose confirm type between danger for delete (that Maria described offline) and default (that we have now) for non destructive actions?

Ron-Lavi commented:
Thanks @ofedoren @ezr-ondrej,
I will add stories and tests too until next week,
and will also work to add the danger confirm modal.

MariaAga commented on webpack/assets/javascripts/react_app/components/ConfirmModal/ConfirmModal.js:
```diff
@@ -0,0 +1,71 @@
+
+ConfirmModal.propTypes = {
+  label: PropTypes.string,
+  message: PropTypes.string,
```
I think message should be a string or a react component in case we want to use links in the modal for example

amirfefer reviewed (commented):
Thanks @laviro , it looks great!

I address here my concerns about redux [here](https://community.theforeman.org/t/adding-confirmation-modal-service-to-the-root-of-our-react-app/24186/3?u=amirfefer)
Is there a way to integrate this service with an `erb` layout?
```ruby
link_to "Dangerous zone", dangerous_zone_path, data: { confirm: 'Are you sure?' }"
```

Ron-Lavi commented:
> Thanks @laviro , it looks great!
> 
> I address here my concerns about redux [here](https://community.theforeman.org/t/adding-confirmation-modal-service-to-the-root-of-our-react-app/24186/3?u=amirfefer)
> Is there a way to integrate this service with an `erb` layout?
> 
> ```ruby
> link_to "Dangerous zone", dangerous_zone_path, data: { confirm: 'Are you sure?' }"
> ```

Thanks @amirfefer I just responded in [discourse](https://community.theforeman.org/t/adding-confirmation-modal-service-to-the-root-of-our-react-app/24186/10?u=laviro),

basically, Rails UJS provides a way to override the confirm modal, explained pretty well in https://derk-jan.com/2020/10/rails-ujs-custom-confirm/

btw Katello implemented it in [common/katello.js#L104](https://github.com/Katello/katello/blob/ea9570ae2e2529b54f20f113abdab86b83371040/app/assets/javascripts/katello/common/katello.js#L104)

EDIT:
added a commit that handles the override, here it is in action:

https://user-images.githubusercontent.com/26363699/124394884-077d5d80-dd0a-11eb-99ee-8c2e3105b73e.mp4

ronlavi2412 committed:
Fixes #32880 - Add confirmation modal service

Ron-Lavi commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -10,6 +10,22 @@ import { sprintf, translate as __ } from './react_app/common/I18n';
+import { openConfirmModal as coreOpenConfirmModal } from './react_app/components/ConfirmModal';
+import { noop } from './react_app/common/helpers';
+
+export function openConfirmModal({
```
now it can be used anywhere, also in legacy code, e.g:
`tfm.tools.openConfirmModal(args)`

btw the original core action is wrapped in this file with `dispatch` so no need to add it.

Ron-Lavi commented on webpack/assets/javascripts/dashboard/index.js:
```diff
@@ -55,29 +59,31 @@ export function removeWidget(item) {
-      },
-    });
-  }
+  openConfirmModal({
+    label: __('Remove widget'),
+    message: __(
+      'Are you sure you want to delete this widget from your dashboard?'
```
should use the warning confirmation modal design here.

ronlavi2412 committed:
Refs #32880 - use confirm modal service

Ron-Lavi commented on webpack/assets/javascripts/dashboard/index.js:
```diff
@@ -55,29 +59,33 @@ export function removeWidget(item) {
-      },
-    });
-  }
+  openConfirmModal({
```
You can test it now in the dashboard as well:

https://user-images.githubusercontent.com/26363699/124386171-cffabb00-dce1-11eb-8413-6320cc0117cc.mp4

ronlavi2412 committed:
Refs #32880 - override rails confirm modal

override the jQuery UJS $.rails.allowAction

ronlavi2412 committed:
Refs #32880 - fix PF4 modal warning

Added the  aria-label prop

Ron-Lavi commented:
The katello code for overriding confirm modal should also be removed for not having duplication and using a unified look for confirm modals

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfirmModal/index.js:
```diff
@@ -0,0 +1,65 @@
+  const closeModal = () => dispatch(closeConfirmModal());
+
+  const handleCancel = () => {
+    onCancel && onCancel();
+    closeModal();
+  };
+
+  const handleConfirm = () => {
+    onConfirm && onConfirm();
```
Looks like these handlers are  initialized with `noop` already, so we won't need to check before calling

Ron-Lavi replied:
true! thanks

amirfefer commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -10,6 +10,59 @@ import { sprintf, translate as __ } from './react_app/common/I18n';
+}
+
+// override the jQuery UJS $.rails.allowAction
+$.rails.allowAction = function railsConfirmOverride(element) {
```
This overrides an inner interface of rails, how about adding an integration test for catching regressions?

Ron-Lavi replied:
do you mean a rails integration test? ok, will try to do something

amirfefer replied:
yes. i.e - an integration test that deletes a resource with a visible confirmation modal

tbrisker replied:
Should we override `$.rails.confirm` instead? looking at the documentation it is ment to be overriden and this method doesn't mention overriding.

tbrisker replied:
I tried overriding `confirm` with an async method but looks like it doesn't work since the default rails method returns ` answer && callback`, and when `answer` is an unresolved promise that is considered a truthy value. 

However, looks like the current implementation doesn't follow the API of the rails implementation, which we should maintain to ensure nothing breaks:
https://github.com/rails/jquery-ujs/blob/master/src/rails.js#L287-L296

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfirmModal/slice.js:
```diff
@@ -0,0 +1,51 @@
+import Immutable from 'seamless-immutable';
```
This slice RFC / suggestion is currently under discussion [here](https://community.theforeman.org/t/rethinking-react-redux-folder-structure/24183), I'd say it worth waiting a bit.

Ron-Lavi replied:
ok, should we wait with the PR until we get a green light from anyone on the discussion or another outcome from it? or should I revert back to `reducer.js, action.js, selectors.js, constants.js`?

amirfefer replied:
I'd suggest keeping the reducers/actions files for now

Ron-Lavi replied:
seem that the RFC got accepted, so let's shift to the new `slice.js` approach IMHO

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfirmModal/slice.js:
```diff
@@ -0,0 +1,51 @@
+import Immutable from 'seamless-immutable';
+import { noop } from '../../common/helpers';
+
+export const OPEN_CONFIRM_MODAL = 'OPEN_CONFIRM_MODAL';
+export const CLOSE_CONFIRM_MODAL = 'CLOSE_CONFIRM_MODAL';
```
do we need to export these constants?

Ron-Lavi replied:
I guess no one will import it really

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfirmModal/slice.js:
```diff
@@ -0,0 +1,51 @@
+  },
+});
+
+export const closeConfirmModal = () => ({
+  type: CLOSE_CONFIRM_MODAL,
+  payload: {},
+});
```
this can be a simple action
```suggestion
export const closeConfirmModal = ({
  type: CLOSE_CONFIRM_MODAL,
  payload: {},
});
```

Ron-Lavi replied:
nice! didn't think about it :)

Ron-Lavi replied:
but then again... I think the convention for actions is a function, and people will try to call it `closeConfirmModal()`.

amirfefer replied:
it's not a convention for actions by redux, just for thunked actions (mainly for asynchronous needs)

amirfefer reviewed (changes requested):
Thanks, @laviro !

Redux makes it much easier with our legacy, no doubt, it would work with a callable approach but I think it doesn't worth the extra effort at the moment. I left a few minor inline comments.

ronlavi2412 committed:
Refs #32880 - Amir's review fixes

Ron-Lavi commented:
Example of using it in the[ index page template draft PR](https://github.com/theforeman/foreman/pull/8596):
![Architectures](https://user-images.githubusercontent.com/26363699/124595768-1b3fd580-de6a-11eb-9970-4744fd96d8fd.gif)

Ron-Lavi commented:
> Example of using it in the[ index page template draft PR](https://github.com/theforeman/foreman/pull/8596):

back to the discussion in https://github.com/theforeman/foreman/pull/8619#discussion_r657935203
if you look carefully, you can see that the confirm modal is being closed, and right after the page reload its data,
I think it might be a good case to let the consumer also the ability to control when to close and to display some loading indicator?
cc @ezr-ondrej

Ron-Lavi commented:
> > Example of using it in the[ index page template draft PR](https://github.com/theforeman/foreman/pull/8596):
> 
> back to the discussion in [#8619 (comment)](https://github.com/theforeman/foreman/pull/8619#discussion_r657935203)
> if you look carefully, you can see that the confirm modal is being closed, and right after the page reload its data,
> I think it might be a good case to let the consumer also the ability to control when to close and to display some loading indicator?

Sorry for all of the comments, maybe... we could show the loading spinner on the table instead.

xprazak2 commented:
> > > Example of using it in the[ index page template draft PR](https://github.com/theforeman/foreman/pull/8596):
> > 
> > 
> > back to the discussion in [#8619 (comment)](https://github.com/theforeman/foreman/pull/8619#discussion_r657935203)
> > if you look carefully, you can see that the confirm modal is being closed, and right after the page reload its data,
> > I think it might be a good case to let the consumer also the ability to control when to close and to display some loading indicator?
> 
> Sorry for all of the comments, maybe... we could show the loading spinner on the table instead.

We could also show the spinner on the modal and close the modal after the async action is finished:
![modal-spinner](https://user-images.githubusercontent.com/2664090/124715853-b6d15480-df03-11eb-8a93-53b09ab61f2a.png)

MariaAga commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -10,6 +10,59 @@ import { sprintf, translate as __ } from './react_app/common/I18n';
+import { noop } from './react_app/common/helpers';
+
+export function openConfirmModal({
+  title = '',
+  message = '',
+  onConfirm = noop,
+  onCancel = noop,
+  modalProps = {},
+  isWarning = false,
+  confirmButtonText = null,
```
The default values are also defined in `coreOpenConfirmModal` should they be also here?

Ron-Lavi replied:
sure, though I will add some comments above the helper for better IntelliSense

tbrisker replied:
I think we should have the defaults set only in one place so we don't need to update two places if they ever change, or we might get inconsistencies between the two (since these defaults will override the component's defaults).
Also, then we wouldn't need to destructure here and can just pass the options object directly to the action.

Ron-Lavi commented:
As for the loading discussion, we decided that the consumer (e.g: page with table)
will handle the loading state, and the modal will just disappear after the confirmation click.
Thanks everyone for sharing your thoughts, and thanks @marisvirik for clarifying the conventions for modals.

tbrisker commented on webpack/assets/javascripts/dashboard/index.js:
```diff
@@ -55,29 +59,33 @@ export function removeWidget(item) {
+    ),
+    confirmButtonText: __('Delete'),
+    isWarning: true,
+    onConfirm: function onConfirm() {
```
could this be an arrow function?

tbrisker commented on webpack/assets/javascripts/react_app/components/ConfirmModal/slice.js:
```diff
@@ -0,0 +1,51 @@
+  }
+};
+
+export const storeDomain = 'confirmModal';
```
now that we have redux toolkit, can you use createSlice instead?

tbrisker commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -10,6 +10,59 @@ import { sprintf, translate as __ } from './react_app/common/I18n';
+  isWarning = false,
+  confirmButtonText = null,
+}) {
+  return store.dispatch(
```
do we need to return the action? 
or could we await and return `true`/`false` depending on the user action somehow, and mimick the `window.confirm` interface (which would also mean if there is a string passed as the argument instead of an object that should be set as the message)? Would that also make overriding the rails confirm much simpler?

tbrisker commented on webpack/assets/javascripts/react_app/components/ConfirmModal/index.js:
```diff
@@ -0,0 +1,65 @@
+      aria-label="application confirm modal"
+      variant={ModalVariant.small}
+      title={title}
+      isOpen={isOpen}
```
Does it make sense to actually render the modal at all if it isn't open? could we return early if it is closed?

tbrisker commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -10,6 +10,59 @@ import { sprintf, translate as __ } from './react_app/common/I18n';
+export function openConfirmModal({
+  title = '',
+  message = '',
+  onConfirm = noop,
+  onCancel = noop,
```
would it make sense to have the default callbacks return true and false? i.e. something like
```suggestion
  onConfirm = () => true,
  onCancel = () => false,
```

ezr-ondrej commented:
What's the status here? Can I help somehow to get this in?

ronlavi2412 committed:
Refs #32880 - Move into a real slice

ronlavi2412 committed:
Refs #32880 - Return null early if not isOpen

ronlavi2412 committed:
Refs #32880 - arrow functions and params refactor

Ron-Lavi commented:
Thanks @tbrisker @ezr-ondrej, update with some of the review comments, will continue hopefully soon :)

tbrisker commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -10,6 +10,39 @@ import { sprintf, translate as __ } from './react_app/common/I18n';
+      title: __('Confirm'),
+      message,
+      onConfirm: () => {
+        callback = $.rails.fire(element, 'confirm:complete', [answer]);
+        if (callback) {
+          const oldAllowAction = $.rails.allowAction;
+          $.rails.allowAction = function allowAction() {
+            return true;
+          };
+          element.trigger('click');
+          $.rails.allowAction = oldAllowAction;
+        }
```
`callback` will always be true here, we don't really have any callbacks on `confirm:complete` anywhere, certainly not ones that return `false`. we can simplify this to:
```suggestion
        const oldAllowAction = $.rails.allowAction;
        $.rails.allowAction = () => true;
        element.trigger('click');
        $.rails.allowAction = oldAllowAction;
```

Ron-Lavi replied:
done, thanks!

tbrisker reviewed (changes requested):
Please open an issue for removing the special handling in Katello so it doesn't conflict with this, and rebase.

ronlavi2412 committed:
Refs #32880 - show as a warning on delete

when using the erb helper for display_delete_if_authorized
show confirm modal in warning styles

ronlavi2412 committed:
Refs #32880 - add integration test

ronlavi2412 committed:
Refs #32880 - add story

Ron-Lavi commented:
opened issue for katello: https://projects.theforeman.org/issues/33393

tbrisker reviewed (dismissed):
Looks great @laviro ! we'll need the katello PR to be ready so we can merge them together.

Ron-Lavi commented:
Opened Katello PR: https://github.com/Katello/katello/pull/9601

ezr-ondrej commented:
Loving it! Can't merge (test) from phone, but having pr for katello unblocks merging this IMHO :)

ronlavi2412 committed:
Refs #32880 - override method only if $.rails exists

Ron-Lavi commented:
> Loving it! Can't merge (test) from phone, but having pr for katello unblocks merging this IMHO :)

Thanks @ezr-ondrej, I think I solved the issue of Katello
cc @Andrewgdewar @sjha4

tbrisker commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -10,6 +10,36 @@ import { sprintf, translate as __ } from './react_app/common/I18n';
+    }
+    return false;
+  };
+}
```
Will this run every time that `foreman_tools` is imported? Perhaps we should move it to a separate `rails_overrides.js` file that is imported once in the bundle?

Ron-Lavi replied:
Moved it to application.js fie's `onContentLoad`
although it was actually called only once each full-refresh in foreman_tools, same as in application.js
it feels that it's a better place to store such overrides :)

Ron-Lavi replied:
moved to `foreman_overrides.js` instead

MariaAga reviewed (dismissed):
Except for Tomers comment, LGTM :)

tbrisker commented on app/assets/javascripts/application.js:
```diff
@@ -87,6 +87,33 @@ function onContentLoad() {
+        title: __('Confirm'),
+        message,
+        isWarning,
+        onConfirm: () => {
```
legacy assets don't go through webpack, so this won't get transpiled by babel

Ron-Lavi replied:
I wish this will be the default in all browsers already,
well, almost there https://caniuse.com/arrow-functions

Ron-Lavi replied:
changed to regular functions

amirfefer commented on app/assets/javascripts/application.js:
```diff
@@ -87,6 +87,32 @@ function onContentLoad() {
+    if (!message) return true;
+
+    if ($.rails.fire(element, 'confirm')) {
+      tfm.tools.openConfirmModal({
```
you can use our [legacy-js store](http://foreman.surge.sh/?path=/docs/introduction-legacy-js--page) and dispatch the action directly:
```js
tfm.store.dispatch(coreOpenConfirmModal(options))
```

Ron-Lavi replied:
Thanks, moved to `foreman_overrides.js` and using it there :)

amirfefer reviewed (commented):
Thanks, @laviro,  looks great! just one concern,
do we want to have an integration test just to make sure that we won't have any regression that might cause by the overridden inner rails code?

ronlavi2412 committed:
Refs #32880 - move rails override to foreman_overrides.js

ronlavi2412 committed:
Refs #32880 - add server side integration test

Ron-Lavi commented:
@amirfefer Added server side integration test :)

Ron-Lavi commented on test/integration/operatingsystem_js_test.rb:
```diff
@@ -4,4 +4,24 @@ class OperatingsystemJSTest < IntegrationTestWithJavascript
+    confirm_button = confirm_modal.find("footer > button:nth-child(1)")
+    assert confirm_button.text, "Confirm"
+    confirm_button.click
+    assert has_no_css?("#app-confirm-modal")
```
@ezr-ondrej @tbrisker @amirfefer maybe you can help with the test?
I need to wait somehow for the browser to reload after the confirm click
and test if the OS was successfuly deleted

tbrisker replied:
Thanks, pushed another commit fixing the test - the problem was that the OS wasn't deleted because it was still in use. Adding `DEBUG_JS_TEST=1` to the test run causes it to run in chrome instead of headless chrome so you can see what it is doing - so I saw it was displaying an error message. Added a new OS with name `aaa` to make sure it is listed first, and also fixed the `assert` that should have been `assert_equal` (`assert` only checks if the value of the first argument is truthy, and if it isn't it displays the second argument as message).

tbrisker committed:
Refs #32880 - Fix integration test

tbrisker commented:
[test katello]

amirfefer reviewed (approved):
LGTM 👍 
Thanks @laviro, @tbrisker !

tbrisker reviewed (approved):
Thanks @laviro ! great to finally have this ready :shipit:
